File: common/src/main/java/dev/number6/comprehend/adaptor/AwsComprehensionAdaptor.java
package dev.number6.comprehend.adaptor;

import dev.number6.comprehend.AwsComprehendClient;
import dev.number6.comprehend.results.PresentableEntityResults;
import dev.number6.comprehend.results.PresentableKeyPhrasesResults;
import dev.number6.comprehend.results.PresentableSentimentResults;
import dev.number6.comprehend.port.ComprehensionPort;
import dev.number6.message.ChannelMessages;
import com.amazonaws.services.comprehend.model.*;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

public class AwsComprehensionAdaptor implements ComprehensionPort {

    private final AwsComprehendClient awsComprehendClient;

    public AwsComprehensionAdaptor(AwsComprehendClient awsComprehendClient) {
        this.awsComprehendClient = awsComprehendClient;
    }

    @Override
    public PresentableEntityResults getEntitiesForSlackMessages(ChannelMessages channelMessages) {
        Collection<DetectEntitiesResult> entityResults = channelMessages.getMessages().stream()
                .filter(Objects::nonNull)
                .filter(m -> m.length() > 0)
                .map(awsComprehendClient::getMessageEntities)
                .collect(Collectors.toList());

        Map<String, Map<String, Long>> entityCollection = entityResults.stream()
                .filter(e -> e.getEntities() != null && e.getEntities().size() > 0)
                .flatMap(e -> e.getEntities().stream())
                .collect(Collectors.groupingBy(Entity::getType, HashMap::new, Collectors.groupingBy(Entity::getText, Collectors.counting())));

        return new PresentableEntityResults(channelMessages.getComprehensionDate(), entityCollection, channelMessages.getChannelName());
    }

    @Override
    public PresentableSentimentResults getSentimentForSlackMessages(ChannelMessages channelMessages) {

        Collection<DetectSentimentResult> sentimentResults = channelMessages.getMessages().stream()
                .filter(Objects::nonNull)
                .filter(m -> m.length() > 0)
                .map(awsComprehendClient::getMessageSentiment)
                .collect(Collectors.toList());

        return new PresentableSentimentResults(channelMessages.getComprehensionDate(),
                sentimentResults,
                channelMessages.getChannelName());
    }

    @Override
    public PresentableKeyPhrasesResults getKeyPhrasesForSlackMessages(ChannelMessages channelMessages) {
        Collection<DetectKeyPhrasesResult> keyPhrasesResults = channelMessages.getMessages().stream()
                .filter(Objects::nonNull)
                .filter(m -> m.length() > 0)
                .map(awsComprehendClient::getMessageKeyPhrases)
                .collect(Collectors.toList());

        Map<String, Long> keyPhrases = keyPhrasesResults.stream()
                .filter(r -> r.getKeyPhrases() != null && r.getKeyPhrases().size() > 0)
                .flatMap(r -> r.getKeyPhrases().stream())
                .collect(Collectors.groupingBy(KeyPhrase::getText, Collectors.counting()));

        return new PresentableKeyPhrasesResults(channelMessages.getComprehensionDate(), keyPhrases, channelMessages.getChannelName());
    }
}

File: common/src/main/kotlin/dev/number6/comprehend/adaptor/AwsComprehensionAdaptor.kt
package dev.number6.comprehend.adaptor

import com.amazonaws.services.comprehend.model.DetectEntitiesResult
import com.amazonaws.services.comprehend.model.DetectKeyPhrasesResult
import com.amazonaws.services.comprehend.model.DetectSentimentResult
import com.amazonaws.services.comprehend.model.KeyPhrase
import dev.number6.comprehend.AwsComprehendClient
import dev.number6.comprehend.port.ComprehensionPort
import dev.number6.comprehend.results.PresentableEntityResults
import dev.number6.comprehend.results.PresentableKeyPhrasesResults
import dev.number6.comprehend.results.PresentableSentimentResults
import dev.number6.message.ChannelMessages
import java.util.*
import java.util.function.Function
import java.util.stream.Collectors

class AwsComprehensionAdaptor(private val awsComprehendClient: AwsComprehendClient) : ComprehensionPort {
    override fun getEntitiesForSlackMessages(channelMessages: ChannelMessages): PresentableEntityResults {
        val entityResults: Collection<DetectEntitiesResult> = channelMessages.messages
                .filter { m -> Objects.nonNull(m) }
                .filter { m -> m?.length!! > 0 }
                .map { message -> awsComprehendClient.getMessageEntities(message) }

        val entityCollection: Map<String, Map<String, Long>> = entityResults
                .filter { e: DetectEntitiesResult -> e.entities != null && e.entities.size > 0 }
                .flatMap { e: DetectEntitiesResult -> e.entities }
                .groupBy { e -> e.type }
                .mapValues { e -> e.value.groupBy { e1 -> e1.text } }
                .mapValues { e -> e.value.mapValues { e2 -> e2.value.size.toLong() } }
//                .collect(Collectors.groupingBy(Function { obj: Entity -> obj.type }, Supplier { HashMap() }, Collectors.groupingBy({ obj: Entity -> obj.text }, Collectors.counting())))
        return PresentableEntityResults(channelMessages.comprehensionDate, entityCollection, channelMessages.channelName)
    }

    override fun getSentimentForSlackMessages(channelMessages: ChannelMessages): PresentableSentimentResults {
        val sentimentResults: Collection<DetectSentimentResult> = channelMessages.messages.stream()
                .filter { obj: String? -> Objects.nonNull(obj) }
                .filter { m: String? -> m!!.length > 0 }
                .map { message: String? -> awsComprehendClient.getMessageSentiment(message) }
                .collect(Collectors.toList())
        return PresentableSentimentResults(channelMessages.comprehensionDate,
                sentimentResults,
                channelMessages.channelName)
    }

    override fun getKeyPhrasesForSlackMessages(channelMessages: ChannelMessages): PresentableKeyPhrasesResults {
        val keyPhrasesResults: Collection<DetectKeyPhrasesResult> = channelMessages.messages.stream()
                .filter { obj: String? -> Objects.nonNull(obj) }
                .filter { m: String? -> m!!.length > 0 }
                .map { message: String? -> awsComprehendClient.getMessageKeyPhrases(message) }
                .collect(Collectors.toList())
        val keyPhrases = keyPhrasesResults.stream()
                .filter { r: DetectKeyPhrasesResult -> r.keyPhrases != null && r.keyPhrases.size > 0 }
                .flatMap { r: DetectKeyPhrasesResult -> r.keyPhrases.stream() }
                .collect(Collectors.groupingBy(Function { obj: KeyPhrase -> obj.text }, Collectors.counting()))
        return PresentableKeyPhrasesResults(channelMessages.comprehensionDate, keyPhrases, channelMessages.channelName)
    }

}